### DISCUSSION TITLE
Anyone else ever considered doing a home built organ pedal midi controller?

### DISCUSSION THREAD

**Discussion by wpostma** [GrandOrgue/grandorgue#1124] [General]
I am contemplating one using the Teensy or low cost arduino mega controllers together.  

I don't think I have room for a "rescued midi capable home organ console" but then, perhaps, given how fun this is, I may some day make space for that.

My current synth system doesn't currently even have room under neath it for some foot pedals but I am thinking if I did make one, it would be smallish just to get the idea of it and then expand it later if I decide I like it.  Perhaps as small as two octaves.

**Comment by thbeckpr** [comment]
Here's a 24 note pedal board on Amazon:
https://www.amazon.com/Tech-Note-Midi-Pedal-Controller/dp/B00TDU2IA8/ref=sr_1_3?crid=2E23XM1O9LRK5&keywords=midi+organ+pedal+board&qid=1652014167&sprefix=midi+organ%2Caps%2C241&sr=8-3

**Reply by davidgritter** [reply]
I have that pedalboard, and it works ok, but is a bit noisy when played, and has a slightly different feel than a real pedalboard.  Has standard midi output only, not usb midi and appears to have an arduino nano as the internal controller.
I've also built arduino midi interfaces.  Using a leonardo, or a teensy, or a raspberry pi pico, direct USB midi is possilble, and the libraries make it simple  Teensy has the ability to name the midi interface, but as far as I can see the other options do not.  If you pick the pico, use the earle pilehower arduino device, as the officially supported MBED device doesn't like it when USB midi is connected to a computer that's still booting up.  The PICO is by far the lowest cost solution at $4.00.   see https://diyorganconsole.wordpress.com for more ideas and some downloadable code, especially if you want to make a complete console at some poing

**Comment by jandeboever** [comment]
I built a flat/baroque pedalboard (30 keys / C to f') myself and then midified it with a kit I bought in the Netherlands. If you're a bit handy with woodworking it is definitely possible. I can share my plans if you want. jan dot deboever at gmail dot com

**Comment by VladimirSlavik** [comment]
I've been considering the same. This is the probably most detailed writeup I found so far: https://fagerjord.org/pipe-organ/524/

**Comment by Aduragbemi-Adeyinka** [comment]
I'm making one presently (as a part of a home-made console). I'm getting lots of help and I'm learning a lot while at it.
Start from here: http://familjenpalo.se/vpo/pedalboard/ 
and if you need more fine details check the site quoted above. Both sites (and more) are immensely helpful for me.

**Comment by sandstone42** [comment]
I had an Allen TC3 pedalboard that I converted to MIDI with the help of an Arduino MEGA board. I had to use a piece of fine emery to clean the contacts but it worked fine at the +5VDC level.

**Comment by William-Purvis** [comment]
I've been working on a Compton Organ - stripped out all the guts and working to replace everything but the case and pedal board.
I've modified the pedals by adding photo sensors, and an scanner based on an Arduino Nano with two 16-bit expanders. Seems to
working well, though we've not reached to point where I can tie all the bits together and get our organist to play it (I'm a guitar/bass player, not an organist! More details available if anyone wants them.

**Comment by djbottrill** [comment]
I've been developing my own midi controllers for a number of years, my latest iteration uses a custom designed printed circuit board and a Raspberry Pi Pico microcontroller. These are daisy-chained together with a 10 way ribbon cable supplying power and an I2C bus for inter-board communications. One board is nominated as a master that can either communicate with the PC using USB Midi or over Ethernet using IP Midi (QMidinet on Linux) this just has an additional Ethernet module soldered onto the board.

Each scanner board, including the master, can support a diode switch key matrix or my custom designed optical matrix boards that can be fitted to some keyboard types. Alternatively the board will support up to 16 LED illuminated push buttons for pistons or stop buttons. Finally each board has 3 analogue inputs for swell controllers or other analogue functions. 

The current system sends Midi 9x midi messages with no velocity for keyboards and NRPN messages for expression pedals or Piston presses it also responds to corresponding NRPN messages to light the piston LEDs.

Theoretically up to 126 boards could be connected but 16 would probably more realistic. I'm currently working on an add-on to drive pipe organ action magnets from received Midi Messages.

Please contact me if anyone is interested in more information / help.

My current setup is 4 manuals, using optical switches, each with 16 pistons and 32 note pedalboard, with reed switches, and 3 swell pedals connected over Ethernet to a Mac Mini running GO on Linux.

**Reply by William-Purvis** [reply]
I'd be interested in exchanging schematics. I'd already designed my scanner before the Pico came out, and with a stock of PCBs
made, there didn't seem to be much point in reworking it. Basically, it's an Arduino Nano mounted on  50x85mm PCB, with two
MCP23S17 16-bit expanders. Output is over the USB port on the Arduino. One thing that I've not been able to figure out is how
to override the identity, and make it look like a MIDI device. To fix that I've written a short C program that opens a number of
ttyUSBxx ports, and feeds the midi events to Jack. I have to sort out a further mod to that as the PC will need to send events
to a specific port, but there's no good way of distinguishing a particular serial port, they can come up as different devices
when you reboot. As each scanner has a pre-programmed unique channel, I can use that, or invent some kind of protocol for the
converter to identify each device on startup. I would also like to dig into the system to be informed if a new scanner is
plugged in after startup, though that's only for debugging.

Here's a rather poor photo of my scanner:
![image](https://user-images.githubusercontent.com/32897843/218251143-0ed7787e-9520-49d5-895e-2a94754c0350.png)
It is actually one of my first batch and there's a bug in the PCB design. As it happens it doesn't matter for the tab scanner, so I've
not remade it. The bug showed up when I tried to use it on the pedal scanner, which needs a  5V supply for the opto sensor,
so that has the revised board.

**Reply by William-Purvis** [reply]
I should point out that the Nano has an awkward size - there are 30 pins. You can only get DIP sockets with 28 or 32 pins.
I used a 32-pin socket with the pair at the right end of the socket unconnected. I have to remind myself which way to plug
in the Nano if I've not done one recently!

**Reply by djbottrill** [reply]
i've just uploaded my Raspberry Pi Pico sketch to my Github https://github.com/djbottrill/RP2040_Midi_Scanner I will upload more information such as circuit diagrams over the coming weeks. Here's a picture of a master scanner on the left and a slave on the right.


![IMG_6856](https://user-images.githubusercontent.com/19348113/218797834-0027edf8-2713-412a-9318-86114b6233ff.jpg)
